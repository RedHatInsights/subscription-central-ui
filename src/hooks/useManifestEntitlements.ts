import { useQuery, QueryObserverResult } from 'react-query';
import Cookies from 'js-cookie';

interface ManifestEntitlement {
  id: string;
  sku: string;
  contractNumber: string;
  entitlementQuantity: number;
  startDate?: string;
  endDate?: string;
  subscriptionName?: string;
}

interface ManifestEntitlementsData {
  body: {
    simpleContentAccess: string;
    createdBy: string;
    createdDate: string;
    entitlementsAttached: EntitlementsAttachedData;
    entitlementsAttachedQuantity: number;
    lastModified: string;
    name: string;
    type: string;
    uuid: string;
    version: string;
  };
}

interface EntitlementsAttachedData {
  valid: boolean;
  reason?: string;
  value?: ManifestEntitlement[];
}

const getManifestEntitlements = (uuid: string): Promise<ManifestEntitlementsData> => {
  const jwtToken = Cookies.get('cs_jwt');
  return fetch(`/api/rhsm/v2/manifests/${uuid}?include=entitlements`, {
    headers: { Authorization: `Bearer ${jwtToken}` }
  }).then((response) => response.json());
};

const useManifestEntitlements = (
  uuid: string
): QueryObserverResult<ManifestEntitlementsData, unknown> => {
  return useQuery<any, Error>(['manifestEntitlements', uuid], () => getManifestEntitlements(uuid));
};

export {
  EntitlementsAttachedData,
  ManifestEntitlementsData,
  ManifestEntitlement,
  useManifestEntitlements as default
};
