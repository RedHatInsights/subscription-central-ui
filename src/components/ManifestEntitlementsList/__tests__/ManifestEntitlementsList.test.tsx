import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import Component, { ManifestEntitlementsListProps } from '../ManifestEntitlementsList';
import { QueryClient, QueryClientProvider } from 'react-query';
import useManifestEntitlements, {
  EntitlementsAttachedData
} from '../../../hooks/useManifestEntitlements';

const queryClient = new QueryClient();
jest.mock('../../../hooks/useManifestEntitlements');

const ManifestEntitlementsList = (props: ManifestEntitlementsListProps) => (
  <QueryClientProvider client={queryClient}>
    <Component {...props} />
  </QueryClientProvider>
);

const mockResponse = (entitlementsAttached: EntitlementsAttachedData, isLoading = false) => {
  (useManifestEntitlements as jest.Mock).mockReturnValue({
    isLoading,
    isSuccess: true,
    isError: false,
    data: {
      body: {
        entitlementsAttached
      }
    }
  });
};

describe('Manifest Entitlements List', () => {
  it('renders correctly', () => {
    const entitlementsData = {
      valid: true,
      value: [
        {
          contractNumber: '12345',
          entitlementQuantity: 10,
          id: 'id123',
          sku: 'sku123',
          startDate: '2021-01-01T00:00:00.000Z',
          endDate: '2022-01-01T00:00:00.000Z'
        }
      ]
    };
    mockResponse(entitlementsData);
    const props = {
      entitlementsRowRef: null as React.MutableRefObject<HTMLSpanElement>,
      uuid: 'abc123'
    };
    const { getByLabelText } = render(<ManifestEntitlementsList {...props} />);

    expect(
      getByLabelText('Manifests table').children[1].firstChild.childNodes[1].textContent
    ).toEqual(entitlementsData.value[0].sku);
  });

  it.skip('fires methods in kebab menu when clicked', () => {
    const consoleSpy = jest.spyOn(console, 'log').mockImplementation();

    const entitlementsData = {
      valid: true,
      value: [
        {
          contractNumber: '12345',
          entitlementQuantity: 10,
          id: 'id123',
          sku: 'sku123',
          startDate: '2021-01-01T00:00:00.000Z',
          endDate: '2022-01-01T00:00:00.000Z'
        }
      ]
    };
    mockResponse(entitlementsData);
    const props = {
      entitlementsRowRef: null as React.MutableRefObject<HTMLSpanElement>,
      uuid: 'abc123'
    };
    const { getByText } = render(<ManifestEntitlementsList {...props} />);

    fireEvent.click(document.querySelector('.pf-c-dropdown__toggle'));
    fireEvent.click(getByText('Remove Subscription'));

    fireEvent.click(document.querySelector('.pf-c-dropdown__toggle'));
    fireEvent.click(getByText('Move Subscription'));

    expect(console.log).toHaveBeenCalledTimes(2);

    consoleSpy.mockRestore();
  });

  it('renders correctly when no entitlements are attached to the manifest', () => {
    mockResponse({
      valid: false,
      reason: 'No Entitlements are attached to the manifest'
    });
    const props = {
      entitlementsRowRef: null as React.MutableRefObject<HTMLSpanElement>,
      uuid: 'abc123'
    };
    const { queryByLabelText } = render(<ManifestEntitlementsList {...props} />);

    expect(queryByLabelText('Manifests table')).toBeNull();
  });

  it('renders with an error message when the API call fails', () => {
    const props = {
      isLoading: false,
      isSuccess: true,
      isError: false,
      entitlementsData: {
        valid: false,
        reason: 'No Entitlements are attached to the manifest'
      },
      entitlementsRowRef: null as React.MutableRefObject<HTMLSpanElement>,
      uuid: 'abc123'
    };
    render(<ManifestEntitlementsList {...props} />);
    expect(document.querySelector('.sub-no-entitlements-reason').firstChild.textContent).toEqual(
      props.entitlementsData.reason
    );
  });

  it('renders loading when loading', () => {
    mockResponse(
      {
        valid: false,
        reason: 'No Manifests found'
      },
      true
    );
    const props = {
      entitlementsRowRef: null as React.MutableRefObject<HTMLSpanElement>,
      uuid: 'abc123'
    };
    const container = render(<ManifestEntitlementsList {...props} />);
    expect(container).toHaveLoader();
  });
});
