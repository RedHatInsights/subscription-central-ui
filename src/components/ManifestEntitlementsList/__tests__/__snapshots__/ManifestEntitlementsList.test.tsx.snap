// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Manifest Entitlements List renders correctly 1`] = `
<div>
  <div>
    <table
      aria-label="Allocations table"
      class="manifests_entitlement-list-table pf-c-table pf-m-grid-md pf-m-no-border-rows"
      data-ouia-component-id="OUIA-Generated-Table-2"
      data-ouia-component-type="PF4/Table"
      data-ouia-safe="true"
      role="grid"
    >
      <thead
        class=""
      >
        <tr
          class=""
          data-ouia-component-id="OUIA-Generated-TableRow-1"
          data-ouia-component-type="PF4/TableRow"
          data-ouia-safe="true"
        >
          <th
            class="pf-m-nowrap"
            data-key="0"
            data-label="Subscription name"
            scope="col"
          >
            Subscription name
          </th>
          <th
            class="pf-m-nowrap"
            data-key="1"
            data-label="SKU"
            scope="col"
          >
            SKU
          </th>
          <th
            class="pf-m-nowrap"
            data-key="2"
            data-label="Contract number"
            scope="col"
          >
            Contract number
          </th>
          <th
            class="pf-m-nowrap"
            data-key="3"
            data-label="Quantity"
            scope="col"
          >
            Quantity
          </th>
          <th
            class="pf-m-nowrap"
            data-key="4"
            data-label="Start date"
            scope="col"
          >
            Start date
          </th>
          <th
            class="pf-m-nowrap"
            data-key="5"
            data-label="End date"
            scope="col"
          >
            End date
          </th>
        </tr>
      </thead>
      <tbody
        class=""
        role="rowgroup"
      >
        <tr
          class=""
          data-ouia-component-id="OUIA-Generated-TableRow-2"
          data-ouia-component-type="PF4/TableRow"
          data-ouia-safe="true"
        >
          <td
            class=""
            data-key="0"
            data-label="Subscription name"
          />
          <td
            class=""
            data-key="1"
            data-label="SKU"
          >
            sku123
          </td>
          <td
            class=""
            data-key="2"
            data-label="Contract number"
          >
            12345
          </td>
          <td
            class=""
            data-key="3"
            data-label="Quantity"
          >
            10
          </td>
          <td
            class=""
            data-key="4"
            data-label="Start date"
          >
            2021-01-01
          </td>
          <td
            class=""
            data-key="5"
            data-label="End date"
          >
            2022-01-01
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Manifest Entitlements List renders correctly when no entitlements are attached to the manifest 1`] = `
<div>
  <div
    class="no-entitlements-reason"
  >
    <p
      tabindex="0"
    >
      No Entitlements are attached to the Allocation
    </p>
  </div>
</div>
`;

exports[`Manifest Entitlements List renders loading when loading 1`] = `
<div>
  <div
    class="entitlement-list-loading-container"
  >
    <div
      class="pf-l-bullseye"
    >
      <span
        aria-valuetext="Loading..."
        class="pf-c-spinner pf-m-xl"
        role="progressbar"
      >
        <span
          class="pf-c-spinner__clipper"
        />
        <span
          class="pf-c-spinner__lead-ball"
        />
        <span
          class="pf-c-spinner__tail-ball"
        />
      </span>
    </div>
  </div>
</div>
`;

exports[`Manifest Entitlements List renders with an error message when the API call fails 1`] = `
<div>
  <div
    class="no-entitlements-reason"
  >
    <p
      tabindex="0"
    >
      No Entitlements are attached to the Allocation
    </p>
  </div>
</div>
`;
